pyvo/discover/image.py

```diff
@@ -0,0 +1,635 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""
+Global image discovery.
+
+This module is intended to cover the much-requested use case "give me all
+images with properties x, y, and z, where these properties for now is the
+location in space, time, and spectrum; sensitivity clearly would be great,
+but that needs standard work.
+
+The code here looks for data in SIA1, SIA2, and Obscore services for now.
+"""
+
+# TODOs:
+# * Evaluate and log Overflow conditions in the three protocols
+# * do more (s_region!) on the basis of the WCS parts in SIA1
+# * In obscore, we probably should make the dataproduct type constraint
+#   configurable (this should really also work for cubes)
+# * Obscore query generation *might* want some extra logic for inclusive,
+#   (as in OR whatever IS NULL) -- but is that a good idea?
+# * It would be nice if we preserved datalink availability and perhaps
+#   even let people do automatic cutouts to the RoI.
+
+import functools
+import threading
+
+import requests
+
+from astropy import units as u
+from astropy import table
+from astropy import time
+
+from ..dam import obscore
+from .. import dal
+from .. import registry
+from ..registry import regtap
+
+
+# imports for type hints
+from typing import Callable, Generator, List, Optional, Set, Tuple
+from astropy.units import quantity
+
+
```

Comment:
If anything else not imported into ``__init__.py`` of the module an exposed as public API from there, it should be added here in ``__all__`` and exposed as such. The rest cannot be referenced in the API docs as a sphinx link.

I would think that a flatter namespace would be nice, so would advocate for exposing anything public though the module, and keep this file as an implementation detail, but I may miss some reasons to do it otherwise. Either case, we need to be explicit.